db/kv_dense_channel_storage_reader.go

```diff
@@ -0,0 +1,221 @@
+				}
+				changes = append(changes, partitionChanges...)
+			}
+		}
+	*/
+	return changes, nil
+}
+
+// Returns PartitionStorageReader for this channel storage reader.  Initializes if needed.
+func (ds *DenseStorageReader) getPartitionStorageReader(partitionNo uint16) (partitionStorage *DensePartitionStorageReader) {
+	var ok bool
+	if partitionStorage, ok = ds.partitionStorage[partitionNo]; !ok {
+		partitionStorage = NewDensePartitionStorageReader(ds.channelName, partitionNo, ds.indexBucket)
+		ds.partitionStorage[partitionNo] = partitionStorage
+	}
+	return partitionStorage
+
+}
+
+// calculateChangedPartitions identifies which vbuckets have changed after sinceClock until toClock, groups these
+// by partition, and returns as a array of PartitionRanges, indexed by partition number.
+func (ds *DenseStorageReader) calculateChanged(sinceClock, toClock base.SequenceClock) (changedVbs []uint16, changedPartitions []*PartitionRange) {
+
+	changedVbs = make([]uint16, 0)
+	changedPartitions = make([]*PartitionRange, ds.partitions.PartitionCount())
+	for vbNoInt, toSeq := range toClock.Value() {
+		vbNo := uint16(vbNoInt)
+		sinceSeq := sinceClock.GetSequence(vbNo)
+		if sinceSeq < toSeq {
+			changedVbs = append(changedVbs, vbNo)
+			partitionNo := ds.partitions.PartitionForVb(vbNo)
+			if changedPartitions[partitionNo] == nil {
+				partitionRange := NewPartitionRange()
+				changedPartitions[partitionNo] = &partitionRange
+			}
+			changedPartitions[partitionNo].SetRange(vbNo, sinceSeq, toSeq)
+		}
+	}
+	return changedVbs, changedPartitions
+}
+
+// DensePartitionStorageReader is a non-caching reader - every read request retrieves the latest from the bucket.
+type DensePartitionStorageReader struct {
+	channelName string      // Channel name
+	partitionNo uint16      // Partition number
+	indexBucket base.Bucket // Index bucket
+}
+
+type PartitionChanges struct {
+	changes map[uint16][]*LogEntry
+}
+
+func NewPartitionChanges() *PartitionChanges {
+	return &PartitionChanges{
+		changes: make(map[uint16][]*LogEntry),
+	}
+}
+func (p *PartitionChanges) GetVbChanges(vbNo uint16) []*LogEntry {
+	if vbchanges, ok := p.changes[vbNo]; ok {
+		return vbchanges
+	} else {
+		return nil
+	}
+}
+
+func (p *PartitionChanges) AddEntry(entry *LogEntry) {
+	_, ok := p.changes[entry.VbNo]
+	if !ok {
+		p.changes[entry.VbNo] = make([]*LogEntry, 0)
+	}
+	p.changes[entry.VbNo] = append(p.changes[entry.VbNo], entry)
+}
+
+func (p *PartitionChanges) Count() int {
+	count := 0
+	for _, vbSet := range p.changes {
+		count += len(vbSet)
+	}
+	return count
+}
+
+func NewDensePartitionStorageReader(channelName string, partitionNo uint16, indexBucket base.Bucket) *DensePartitionStorageReader {
+	storage := &DensePartitionStorageReader{
+		channelName: channelName,
+		partitionNo: partitionNo,
+		indexBucket: indexBucket,
+	}
+	return storage
+}
+
+func (r *DensePartitionStorageReader) GetChanges(partitionRange PartitionRange) (*PartitionChanges, error) {
```

Comment:
This also looks like a good candidate for a unit test if there isn't already one